From 50737d2b9acab47cc0d581c6f89a3b62bd70a126 Mon Sep 17 00:00:00 2001
From: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>
Date: Tue, 5 May 2020 23:11:29 -0700
Subject: [PATCH] Expose PMDPort socket memory allocation

This now allows to control socket from which memory for port's queue
setup comes from. Default behavior unchanged.

p = PMDPort(pci='0000:4b:00.0', socket_id=0)

Signed-off-by: Saikrishna Edupuganti <saikrishna.edupuganti@intel.com>
---
 core/drivers/pmd.cc           | 12 +++++++++---
 protobuf/ports/port_msg.proto |  3 +++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/core/drivers/pmd.cc b/core/drivers/pmd.cc
index a47cdd28..68b299cf 100644
--- a/core/drivers/pmd.cc
+++ b/core/drivers/pmd.cc
@@ -250,10 +250,15 @@ CommandResponse PMDPort::Init(const bess::pb::PMDPortArg &arg) {
     return CommandFailure(-ret, "rte_eth_dev_configure() failed");
   }
 
-  int sid = rte_eth_dev_socket_id(ret_port_id);
+  int sid = arg.socket_case() == bess::pb::PMDPortArg::kSocketId ?
+	  arg.socket_id() : rte_eth_dev_socket_id(ret_port_id);
+  /* if soocket_id is invalid, set to 0 */
   if (sid < 0 || sid > RTE_MAX_NUMA_NODES) {
-    sid = 0;  // if socket_id is invalid, set to 0
+    LOG(WARNING) << "Invalid socket, falling back... ";
+    sid = 0;
   }
+  LOG(INFO) << "Initializing Port:" << ret_port_id
+	    << " with memory from socket " << sid;
 
   eth_rxconf = dev_info.default_rxconf;
   eth_rxconf.rx_drop_en = 1;
@@ -293,7 +298,8 @@ CommandResponse PMDPort::Init(const bess::pb::PMDPortArg &arg) {
   }
   dpdk_port_id_ = ret_port_id;
 
-  int numa_node = rte_eth_dev_socket_id(static_cast<int>(ret_port_id));
+  int numa_node = arg.socket_case() == bess::pb::PMDPortArg::kSocketId ?
+              sid : rte_eth_dev_socket_id(ret_port_id);
   node_placement_ =
       numa_node == -1 ? UNCONSTRAINED_SOCKET : (1ull << numa_node);
 
diff --git a/protobuf/ports/port_msg.proto b/protobuf/ports/port_msg.proto
index ff6e65ed..1e8b6af8 100644
--- a/protobuf/ports/port_msg.proto
+++ b/protobuf/ports/port_msg.proto
@@ -47,6 +47,9 @@ message PMDPortArg {
   bool vlan_offload_rx_strip = 5;
   bool vlan_offload_rx_filter = 6;
   bool vlan_offload_rx_qinq = 7;
+  oneof socket {
+    int32 socket_id = 8;
+  }
 }
 
 message UnixSocketPortArg {
-- 
2.25.1

